File: scripts/check_markdown_scripts.py
#!/usr/bin/env python3
"""Reject long executable shell fences in Markdown."""

import argparse
from dataclasses import dataclass
from pathlib import Path

SHELL_LANGUAGES = frozenset({"bash", "sh", "shell", "zsh"})
MAX_SCRIPT_LINES = 10


@dataclass(frozen=True, slots=True)
class Violation:
    path: Path
    line: int
    language: str
    lines: int


def violations(path: Path, /) -> list[Violation]:
    found: list[Violation] = []
    fence_marker = ""
    fence_length = 0
    language = ""
    start = 0
    count = 0
    for line_number, line in enumerate(path.read_text().splitlines(), 1):
        stripped = line.lstrip(" ")
        indentation = len(line) - len(stripped)
        marker = stripped[:1]
        marker_length = len(stripped) - len(stripped.lstrip(marker)) if marker else 0
        if (
            not fence_marker
            and indentation <= 3
            and marker in {"`", "~"}
            and marker_length >= 3
        ):
            fence_marker = marker
            fence_length = marker_length
            info = stripped[marker_length:].strip()
            language = info.split(maxsplit=1)[0].lower() if info else ""
            start = line_number
            count = 0
        elif (
            fence_marker
            and indentation <= 3
            and marker == fence_marker
            and marker_length >= fence_length
            and not stripped[marker_length:].strip(" \t")
        ):
            if language in SHELL_LANGUAGES and count > MAX_SCRIPT_LINES:
                found.append(Violation(path, start, language, count))
            fence_marker = ""
        elif fence_marker:
            count += 1
    if fence_marker and language in SHELL_LANGUAGES and count > MAX_SCRIPT_LINES:
        found.append(Violation(path, start, language, count))
    return found


def markdown_files(paths: list[Path], /) -> list[Path]:
    files: set[Path] = set()
    for path in paths:
        if path.is_dir():
            files.update(path.rglob("*.md"))
        elif path.suffix == ".md":
            files.add(path)
    return sorted(files)


def main() -> int:
    parser = argparse.ArgumentParser()
    parser.add_argument("paths", nargs="+", type=Path)
    args = parser.parse_args()
    found = [item for path in markdown_files(args.paths) for item in violations(path)]
    for item in found:
        print(f"{item.path}:{item.line}: {item.language} fence has {item.lines} lines")
    return bool(found)


if __name__ == "__main__":
    raise SystemExit(main())

File: scripts/check_markdown_scripts.ts
#!/usr/bin/env bun
import { readdirSync, readFileSync, statSync } from "node:fs";
import { extname, join } from "node:path";

export const SHELL_LANGUAGES = new Set(["bash", "sh", "shell", "zsh"]);
export const MAX_SCRIPT_LINES = 10;

export interface Violation {
  readonly path: string;
  readonly line: number;
  readonly language: string;
  readonly lines: number;
}

export function violations(path: string): Violation[] {
  const found: Violation[] = [];
  let fenceMarker = "";
  let fenceLength = 0;
  let language = "";
  let start = 0;
  let count = 0;
  const content = readFileSync(path, "utf8");
  const lines = content.split(/\r\n|\r|\n/);
  if (/\r\n$|[\r\n]$/.test(content)) lines.pop();
  for (const [lineIndex, line] of lines.entries()) {
    const stripped = line.replace(/^ +/, "");
    const indentation = line.length - stripped.length;
    const marker = stripped.slice(0, 1);
    let markerLength = 0;
    while (marker && stripped[markerLength] === marker) markerLength += 1;
    if (!fenceMarker && indentation <= 3 && (marker === "`" || marker === "~") && markerLength >= 3) {
      fenceMarker = marker;
      fenceLength = markerLength;
      language = stripped.slice(markerLength).trim().split(/\s+/, 1)[0]?.toLowerCase() ?? "";
      start = lineIndex + 1;
      count = 0;
    } else if (fenceMarker && indentation <= 3 && marker === fenceMarker && markerLength >= fenceLength && stripped.slice(markerLength).trim() === "") {
      if (SHELL_LANGUAGES.has(language) && count > MAX_SCRIPT_LINES) found.push({ path, line: start, language, lines: count });
      fenceMarker = "";
    } else if (fenceMarker) {
      count += 1;
    }
  }
  if (fenceMarker && SHELL_LANGUAGES.has(language) && count > MAX_SCRIPT_LINES) found.push({ path, line: start, language, lines: count });
  return found;
}

export function markdownFiles(paths: readonly string[]): string[] {
  const files = new Set<string>();
  const visit = (path: string): void => {
    const status = statSync(path);
    if (status.isDirectory()) {
      for (const name of readdirSync(path)) visit(join(path, name));
    } else if (extname(path) === ".md") files.add(path);
  };
  for (const path of paths) visit(path);
  return [...files].sort();
}

export function main(args = Bun.argv.slice(2)): number {
  if (args.length === 0) {
    process.stderr.write("usage: check_markdown_scripts.ts paths [paths ...]\n");
    return 2;
  }
  const found = markdownFiles(args).flatMap(violations);
  for (const item of found) process.stdout.write(`${item.path}:${item.line}: ${item.language} fence has ${item.lines} lines\n`);
  return found.length > 0 ? 1 : 0;
}

if (import.meta.main) process.exit(main());
